--- trunk/modules/mappers/mod_dir.c	2015/04/21 12:21:10	1675102
+++ trunk/modules/mappers/mod_dir.c	2015/04/21 12:23:33	1675103
@@ -387,8 +387,15 @@
 static int dir_fixups(request_rec *r)
 {
     if (r->finfo.filetype == APR_DIR) {
-        /* serve up a directory */
-        return fixup_dir(r);
+        if (fixup_dir(r) == OK) {
+            return OK;
+        }
+        /* we're running between mod_rewrites fixup and its internal redirect handler, step aside */
+        if (!strcmp(r->handler, "redirect-handler")) {
+            return DECLINED;
+        }
+
+        return fixup_dflt(r);
     }
     else if ((r->finfo.filetype == APR_NOFILE) && (r->handler == NULL)) {
         /* No handler and nothing in the filesystem - use fallback */
