
Override default.

--- httpd-2.4.33/modules/md/mod_md_config.c.mddefault
+++ httpd-2.4.33/modules/md/mod_md_config.c
@@ -54,10 +54,14 @@
 
 #define DEF_VAL     (-1)
 
+#ifndef MD_DEFAULT_STORE_DIR
+#define MD_DEFAULT_STORE_DIR "state/md"
+#endif
+
 /* Default settings for the global conf */
 static md_mod_conf_t defmc = {
     NULL,
-    "md",
+    MD_DEFAULT_STORE_DIR,
     NULL,
     NULL,
     80,
